core/vdbe/mod.rs

```diff
@@ -3086,22 +3085,26 @@ impl Program {
         tracing::trace!("Halt auto_commit {}", auto_commit);
         if auto_commit {
             let current_state = connection.transaction_state.borrow().clone();
-            if current_state == TransactionState::Read {
+            if matches!(
+                current_state,
+                TransactionState::Read | TransactionState::Write
+            ) {
                 pager.end_read_tx()?;
```

Comment:
Quick drive-by comment: does it not matter that `pager.end_read_tx()` might be called multiple times for `TransactionState::Write` in case of `CheckpointStatus::IO`?

Reply:
Does matter! I guess I didn't pay attention right now. Will fix it

Reply:
Fixed!